packages/mock-block-dock/src/datastore/use-mock-datastore.ts

```diff
@@ -0,0 +1,618 @@
+              },
+            ],
+          };
+        }
+        const entityId = uuid();
+        const { entityTypeId, properties, linkData } = data;
+        const newEntity: Entity = {
+          metadata: {
+            editionId: {
+              baseId: entityId,
+              versionId: new Date().toISOString(),
+            },
+            entityTypeId,
+          },
+          properties,
+          linkData,
+        };
+
+        setGraph((currentGraph) => {
+          // A shallow copy should be enough to trigger a re-render
+          const newSubgraph = {
+            ...currentGraph,
+          };
+          addEntitiesToSubgraph(newSubgraph, [newEntity]);
+          return newSubgraph;
+        });
+        return { data: newEntity };
+      },
+      [readonly, setGraph],
+    );
+
+  const getEntity: EmbedderGraphMessageCallbacks["getEntity"] = useCallback(
+    async ({ data }) => {
+      if (!data) {
+        return {
+          errors: [
+            {
+              code: "INVALID_INPUT",
+              message: "getEntity requires 'data' input",
+            },
+          ],
+        };
+      }
+
+      const entitySubgraph = getEntityImpl(data, graph);
+
+      if (!entitySubgraph) {
+        return {
+          errors: [
+            {
+              code: "NOT_FOUND",
+              message: `Could not find entity with entityId '${data.entityId}'`,
+            },
+          ],
+        };
+      }
+      return { data: entitySubgraph };
+    },
+    [graph],
+  );
+
+  const updateEntity: EmbedderGraphMessageCallbacks["updateEntity"] =
+    useCallback(
+      async ({ data }) => {
+        if (readonly) {
+          return readonlyErrorReturn;
+        }
+
+        if (!data) {
+          return {
+            errors: [
+              {
+                code: "INVALID_INPUT",
+                message: "updateEntity requires 'data' input",
+              },
+            ],
+          };
+        }
+
+        return new Promise((resolve) => {
+          setGraph((currentGraph) => {
+            const {
+              entityId,
+              entityTypeId,
+              properties,
+              leftToRightOrder,
+              rightToLeftOrder,
+            } = data;
+            const currentEntity = getEntityFromSubgraph(currentGraph, entityId);
+
+            if (currentEntity === undefined) {
+              resolve({
+                errors: [
+                  {
+                    code: "NOT_FOUND",
+                    message: `Could not find entity with entityId '${data.entityId}'`,
+                  },
+                ],
+              });
+              return currentGraph;
+            }
+
+            /**
+             * @todo - This assumes that once an entity is a link, it's always a link, and its endpoints can't be changed
+             *   however we don't enforce this currently. We should probably check the given entityType here
+             */
```

Comment:
This has some wider implications about the endpoints being mutable, which we were thinking they shouldn't be